Adopt qa_automation to QAM maintenance update test with result reporting simplified

tonyyuan1 opened the pull request:
Adopt qa_automation to QAM maintenance update test with test result reporting simplified.

This execution is entirely based on qa_automation/qa_run. The test result is simplfied to only show failed testcases
triggered by regression bug. That will satisfy the need for maintenance update test
Using YAML_SCHEDULE to schedule is recommanded to keep flexibility.
qaset_pre_patch_run, patch_and_reboot and qaset_post_patch_run should schedule in order.
- Related ticket: progress.opensuse.org/issues/63739
- Needles: no
- Verification run: 
sles12sp4: http://10.67.17.201/tests/507
sles12sp3: http://10.67.17.201/tests/512
sles12sp2: http://10.67.17.201/tests/518
sles15: http://10.67.17.201/tests/513
sles15sp1: http://10.67.17.201/tests/523

tonyyuan1 commented:
> Seems your git commit message was mangled as it only shows "result of qa_aotumation" and not the description that you have in the PR description. Maybe wrong keys pressed in vim? Happens to me from time to time ;)

Hello Oliver,
Thank you for reviewing.  I have fixed the commit message.  
One question about the Check:
The code of this PR is on top of qa_automation/qa_run. It relies on test result of qa_run.  Can I use qa_run as parent class to remove duplicate code in this PR to pass the CHECK?  I noticed that qa_run is test module rather than a lib module. I am not sure if It can't be inherited by openQA rule.

okurz commented:
> I am not sure if It can't be inherited by openQA rule.

I am sorry, I don't understand this sentence. If you mean if you can use qa_run as a base class even though it is a test module: Yes, you can do that. Other test modules use the same to reuse common code.

tonyyuan1 commented:
Yes, it's what I mean. I have modified the code.

okurz commented on tests/qa_automation/qaset_post_patch_run.pm:
```diff
@@ -0,0 +1,148 @@
+    # JUnit xml report
+    assert_script_run("/usr/share/qa/qaset/bin/junit_xml_gen.py -n 'regression' -d -o /tmp/junit.xml /var/log/qaset");
+
+    # Record into failure for known bugs to aviod triggering whole test failure.
```
s/aviod/avoid/

tonyyuan1 replied:
corrected.

okurz commented on tests/qa_automation/qaset_post_patch_run.pm:
```diff
@@ -0,0 +1,148 @@
+
+    unless ($testrun_finished) {
+        die "Test run didn't finish within time limit";
+    }
```
you can do just

```
die 'Test run didn't finish within time limit' unless $testrun_finished;
```

tonyyuan1 replied:
Modified.

okurz commented on tests/qa_automation/qaset_post_patch_run.pm:
```diff
@@ -0,0 +1,148 @@
+        die "Test run didn't finish within time limit";
+    }
+
+    # Switch back to root-console
```
I guess this comment is obvious and not necessary :) Same in line 137

tonyyuan1 replied:
Deleted it.

okurz commented on tests/qa_automation/qaset_post_patch_run.pm:
```diff
@@ -0,0 +1,148 @@
+# SUSE's openQA tests
+#
+# Copyright © 2009-2013 Bernhard M. Wiedemann
+# Copyright © 2012-2019 SUSE LLC
```
As this is a new file please just use the current year(s), e.g. `2019-2020 SUSE LLC` or `2020 SUSE LLC` depending on when you worked on this.

tonyyuan1 replied:
Fixed.

okurz commented on tests/qa_automation/qaset_pre_patch_run.pm:
```diff
@@ -0,0 +1,93 @@
+# SUSE's openQA tests
+#
+# Copyright © 2009-2013 Bernhard M. Wiedemann
+# Copyright © 2012-2019 SUSE LLC
```
same as above

tonyyuan1 replied:
Fixed.

okurz commented on tests/qa_automation/qaset_pre_patch_run.pm:
```diff
@@ -0,0 +1,93 @@
+    assert_script_run("$cmd");
+    unless ($testrun_finished) {
+        die "Test run didn't finish within time limit";
+    }
```
same as above

tonyyuan1 replied:
Modified.

okurz commented on tests/qa_automation/qaset_post_patch_run.pm:
```diff
@@ -0,0 +1,140 @@
+
+# Create qaset/config file, reset qaset, and start testrun
+sub start_testrun {
+    my $self = shift;
```
you are not using `$self` so I suggest to just define it as self-standing function, i.e. delete this line and not call as `$self->start_testrun()` but just `start_testrun()`

tonyyuan1 replied:
fixed.

okurz commented:
Ok, all the added code looks fine but I don't see it called anywhere and also no other code removed. Is this intended to replace existing code? If yes, please remove the corresponding code in the same commit(s). If this is for *new* test scenarios then you should add the test modules accordingly to test schedules.

okurz commented:
Logical changes look good. Failed travis CI tests, maybe you are running an old perltidy version?

tonyyuan1 commented:
> Ok, all the added code looks fine but I don't see it called anywhere and also no other code removed. Is this intended to replace existing code? If yes, please remove the corresponding code in the same commit(s). If this is for _new_ test scenarios then you should add the test modules accordingly to test schedules.
This is for _new_ test scenarios. I added it in "load_consoletests". 
Actually, I prefer to schedule it with  YAML_SCHEDULE  over main.pm as load_consoletests has already loaded a lot of tests and all of them are not external harness tests. 
I would like it to be scheduled like grub2_lvm in  Maintenance: single Incidents as follows
https://openqa.suse.de/tests/overview?distri=sle&version=15&build=%3A14217%3Aopenssl-1_1&groupid=159 

Any advice?

okurz commented:
My advice would be to introduce this change wherever tests are existing based on "qa_automation". You can also talk to the job group and test scenario responsibles.

tonyyuan1 commented:
> My advice would be to introduce this change wherever tests are existing based on "qa_automation". You can also talk to the job group and test scenario responsibles.

No, the change can't be introduced to qa_automation. This two test modules  can only be used to test regression for maintenance updates/patches (to test two versions of the same packages to find out whether a regression is introduced between them. )
More info about at https://confluence.suse.com/display/~tonyyuan/2020/01/07/Enhancement+for+test+result+of+qa_automation+for+update+test

I think,  to test  the two versions of same packages in two Builds need two openQA chained jobs. This can't do that. This two modules must run in one job.

okurz commented:
Now I am confused. Your change is called "… for qa_automation" but you say this can not be "introduced to qa_automation". But I don't have that much active experience in this special field. This is why I suggested to involve the corresponding job group and test scenario responsibles here.

tonyyuan1 commented:
> Now I am confused. Your change is called "… for qa_automation" but you say this can not be "introduced to qa_automation". But I don't have that much active experience in this special field. This is why I suggested to involve the corresponding job group and test scenario responsibles here.

I am sorry. The title indeed was inaccurate and misleading. I have changed it so that people could better understand this PR.

okurz commented:
one typo s/maintaince/maintenance/

I understand it better now. But still, I suggested to involve the corresponding job group and test scenario responsibles here. It could very well be that some of the other solutions are not used anywhere else so I don't see it as beneficial to add another way without removing the older, unused parts.

@Soulofdestiny at least from QSF-u I know that you reduced the use of "qa_automation", maybe even use none at all anymore. Can you please also share your current status?

tonyyuan1 commented:
The PR can run any testsuite in user_space(qa_test_*). It run twice, once before update/patch install, the other after it.
It cpmpares the two results and filters PASSED and BROKEN testcases. Only the testcases passed before and failed after are kept. 
Their failure are triggerd by regression bugs. 
Schedule the test in Maintenance: Single Incidents and Maintenance: Test Repo
1. create a testsuite "mau-qa-userspace_all" which will yaml schedule : schedule/qam/%VERSION%/mau-userspace.yaml

2. Job group: "Maintenance: SLE 12 SP5 Incidents"
scenarios:
  x86_64:
     sle-12-SP5-Server-DVD-Incidents-x86_64:
    - mru-install-minimal-with-addons
    - mru-install-minimal-with-addons-variants:
         testsuite: mru-install-minimal-with-addons
         settings:
           USER_SPACE_TESTSUITES=bash,gzip,cpio,openssl
    - mau-qa-userspace_all

Maintenance: Test Repo
scenarios:
  x86_64:
    sle-15-Server-DVD-Updates-x86_64:
    - mru-install-minimal-with-addons
    - mru-install-minimal-with-addons-variants:
         testsuite: mru-install-minimal-with-addons
         settings:
           USER_SPACE_TESTSUITES=bash,gzip,cpio,openssl
    - mau-qa-userspace_all

USER_SPACE_TESTSUITES variable is used to specify which testsuite to run .i.e USER_SPACE_TESTSUITES=bash,gzip,cpio,openssl.

USER_SPACE_TESTSUITES also delay updates install in mru-install-minimal-with-addons until mau-qa-userspace_all (main_common.pm)

tonyyuan1 commented:
@mimi1vx Could you please review if the PR or test scenario is possible ?

tonyyuan1 commented:
@okurz I modified the PR and added comment about where to schedule and how to schedule it.  Ondrej is responsible for  test scenarios of job groups (Maintenance: Single Incidents and Maintenance: Test Repo) for QAM.  He has reviewed it and approved it. Could you please review this PR again?
